P4 device

pbhide opened the pull request:
Trying (playing with) pull requests

Commit:
Add P4 support to rocker device

Commit:
Add b-model libraries to git for now... these will be removed later once all environments are integrated

Commit:
Add copyright messages
Add world name to QMP command
(qemu) info rocker-ports sw1
            ena/    speed/ auto    world
      port  link    duplex neg?    name
     sw1p1  up     10G  FD  No     OF_DPA
     sw1p2  up     10G  FD  No     OF_DPA
     sw1p3  up     10G  FD  No     p4_l2l3
     sw1p4  up     10G  FD  No     p4_l2l3

Commit:
cosmetic changes

Commit:
fix make clean to not remove p4 libs
add new show cmd parameter to json file

Commit:
Remove the need for global port array by providing context to tx_callback function

Commit:
Take care of comments from Scott (except for a few where I have questions)
Take care of checkpatch.pl errors (except for generated files and switch-case
indentation in rocker_p4_rmt(or l2l3).c

scottfeldman commented on hw/net/rocker/rocker.c:
```diff
@@ -435,6 +438,7 @@ static int cmd_consume(Rocker *r, DescInfo *info)
     cmd = rocker_tlv_get_le16(tlvs[ROCKER_TLV_CMD_TYPE]);
     info_tlv = tlvs[ROCKER_TLV_CMD_INFO];
 
+
```
added whitespce

scottfeldman commented on hw/net/rocker/rocker_hw.h:
```diff
@@ -146,6 +147,7 @@ typedef struct rocker_tlv {
 enum {
     ROCKER_TLV_CMD_UNSPEC,
     ROCKER_TLV_CMD_TYPE,                /* u16 */
+    ROCKER_TLV_CMD_WORLD,               /* u32 - world id */
```
Since the ROCKER_TLV_CMD_TYPE_P4_xxx cmds are already only for P4 world, it seems we don't need this ROCKER_TLV_CMD_WORLD.  In other words, the world is encoded in the cmd type.

pbhide replied:
Yes, currently the cmds are unique across the worlds (but don't have to
be). I started without the world_id tlv but introduced the world_id tlv for
the following two reasons. I also took the cue from the comment just above
the switch in cmd_consume() function. -
1. When a new cmd for any world is added, I had to add it in two places the
   switch in cmd_consume() function in rocker.c and again in world-specific
   command handler (e.g of_dpa_cmd() or rocker_p4_rmt_cmd())
2. As we add more P4 pipelines (worlds), we plan to provide an uniform
   interface for all table operations (CRUD). So new table specific TLVs are
   not visible in rocker_rmt_cmd either).

Basically, when something changes in a world (pipeline), I wanted to
minimize code changes to the common code in rocker.c and rocker_p4_rmt.c
(this file is common to different P4 worlds). With world_id tlv, I removed
changes in rocker.c and with a table_op blob, I do not need to change the
rocer_p4_rmt.c
Also as we add more worlds they can have there own TLV space (like nested
TLVs, but w/o adding another level of nesting).

On Sun, Apr 19, 2015 at 9:03 PM, Scott Feldman notifications@github.com
wrote:

> In hw/net/rocker/rocker_hw.h
> https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28663018:
> 
> > @@ -146,6 +147,7 @@ typedef struct rocker_tlv {
> >  enum {
> >      ROCKER_TLV_CMD_UNSPEC,
> >      ROCKER_TLV_CMD_TYPE,                /\* u16 */
> > -    ROCKER_TLV_CMD_WORLD,               /\* u32 - world id */
> 
> Since the ROCKER_TLV_CMD_TYPE_P4_xxx cmds are already only for P4 world,
> it seems we don't need this ROCKER_TLV_CMD_WORLD. In other words, the world
> is encoded in the cmd type.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pbhide/qemu-rocker/pull/1/files#r28663018.

scottfeldman replied:
Ok, those reasons make sense.  I guess OF-DPA is the exception.  We might be able to promote of-dpa to this model without breaking backward-compat with older drivers by assuming if ROCKER_TLV_CMD_WORLD is not defined in the tlvs then it must be an of-dpa cmd.

scottfeldman commented on hw/net/rocker/rocker_hw.h:
```diff
@@ -165,6 +167,14 @@ enum {
     ROCKER_TLV_CMD_TYPE_OF_DPA_GROUP_DEL,
     ROCKER_TLV_CMD_TYPE_OF_DPA_GROUP_GET_STATS,
 
+    ROCKER_TLV_CMD_TYPE_CLEAR_PORT_STATS,
```
Are these STATS updates from David Ahern?  I know he's been patiently waiting to get some rocker device stats patches in.  I should just pull them in and update scottfeldman/qemu-rocker.

pbhide replied:
Yes, I copied them from the driver code when I updated net-next. Ran into
tlv types not matching between driver and device. I think once all repos
are up to date this will not show up as a diff.

On Sun, Apr 19, 2015 at 9:08 PM, Scott Feldman notifications@github.com
wrote:

> In hw/net/rocker/rocker_hw.h
> https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28663127:
> 
> > @@ -165,6 +167,14 @@ enum {
> >      ROCKER_TLV_CMD_TYPE_OF_DPA_GROUP_DEL,
> >      ROCKER_TLV_CMD_TYPE_OF_DPA_GROUP_GET_STATS,
> > -    ROCKER_TLV_CMD_TYPE_CLEAR_PORT_STATS,
> 
> Are these STATS updates from David Ahern? I know he's been patiently
> waiting to get some rocker device stats patches in. I should just pull them
> in and update scottfeldman/qemu-rocker.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pbhide/qemu-rocker/pull/1/files#r28663127.

scottfeldman replied:
makes sense

scottfeldman commented on hw/net/rocker/rocker_hw.h:
```diff
@@ -165,6 +167,14 @@ enum {
+    ROCKER_TLV_CMD_TYPE_CLEAR_PORT_STATS,
+    ROCKER_TLV_CMD_TYPE_GET_PORT_STATS,
+
+    ROCKER_TLV_CMD_TYPE_P4_RMT_TABLE_ENTRY_ADD,
```
What does the "RMT" part of "P4_RMT" mean?

pbhide replied:
RMT = Reconfigurable Match Table  - This achronym comes from one of the
initial papers on reconfigurable packet processing (I'll send you the exact
name of the paper).
I am using it here as a common identifier that can be shared by P4 worlds
(L2L3 and future). Each P4 world will add its own tables but share a single
TLV type to ADD and table entry.

On Sun, Apr 19, 2015 at 9:09 PM, Scott Feldman notifications@github.com
wrote:

> In hw/net/rocker/rocker_hw.h
> https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28663140:
> 
> > @@ -165,6 +167,14 @@ enum {
> >      ROCKER_TLV_CMD_TYPE_OF_DPA_GROUP_DEL,
> >      ROCKER_TLV_CMD_TYPE_OF_DPA_GROUP_GET_STATS,
> > -    ROCKER_TLV_CMD_TYPE_CLEAR_PORT_STATS,
> > -    ROCKER_TLV_CMD_TYPE_GET_PORT_STATS,
> >   +
> > -    ROCKER_TLV_CMD_TYPE_P4_RMT_TABLE_ENTRY_ADD,
> 
> What does the "RMT" part of "P4_RMT" mean?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pbhide/qemu-rocker/pull/1/files#r28663140.

scottfeldman replied:
I see.  Maybe there is a place for a comment in the code that expands the acronym?

scottfeldman commented on hw/net/rocker/rocker_hw.h:
```diff
@@ -475,6 +504,16 @@ enum rocker_of_dpa_overlay_type {
 
+/* cmd info (nested tlvs) for P4 */
+enum {
+    ROCKER_TLV_P4_RMT_INFO_UNSPEC = 0,  /* don't use 0 for tlv type */
```
Can probably drop the "INFO" part of the name to keep things shorter and be consistent with OF-DPA cmd info tlvs.

pbhide replied:
I wanted to have something in TLV that tells me that this is a nested TLV
(so it won't be used incorrectly).
I think, I wanted to drop INFO and keep CMD but did the reverse by mistake.
E.g. I wanted -
ROCKER_TLV_P4_RMT_CMD_UNSPEC, ..
This will be consistent with -
ROCKER_TLV_CMD_PORT_SETTINGS_UNSPEC

Eitherways.. it really does not shorten the name. I can drop both INFO and
CMD (altough prefer to keep some indication that this is nested tlv under
CMD)

On Sun, Apr 19, 2015 at 9:12 PM, Scott Feldman notifications@github.com
wrote:

> In hw/net/rocker/rocker_hw.h
> https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28663229:
> 
> > @@ -475,6 +504,16 @@ enum rocker_of_dpa_overlay_type {
> >      (ROCKER_GROUP_TYPE_SET(ROCKER_OF_DPA_GROUP_TYPE_L3_UCAST) |\
> >       ROCKER_GROUP_INDEX_LONG_SET(index))
> > 
> > +/\* cmd info (nested tlvs) for P4 */
> > +enum {
> > -    ROCKER_TLV_P4_RMT_INFO_UNSPEC = 0,  /\* don't use 0 for tlv type */
> 
> Can probably drop the "INFO" part of the name to keep things shorter and
> be consistent with OF-DPA cmd info tlvs.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pbhide/qemu-rocker/pull/1/files#r28663229.

scottfeldman replied:
Ok, maybe INFO is best since that's the nest it's under.  Again, of-dpa will be an exception, but maybe we can fix of-dpa to be consistent once the dust settles.

scottfeldman commented on hw/net/rocker/rocker_world.c:
```diff
@@ -97,10 +108,8 @@ enum rocker_world_type world_type(World *world)
-        return "OF_DPA";
-    default:
-        return "unknown";
+    if (world) {
```
Shouldn't need to check if world is non-null here.  If NULL was passed in, that's a bug in the calling code and I think we'd rather have the device crash (so the bug can get fixed) rather than returning something that sort of look like world is non-NULL.

scottfeldman replied:
I would say just return world->name here and that's it.

pbhide replied:
Agree. I really wanted as assert.. but don't see it being used anywhere.
Also wanted to somewhat mimic the default: case in the earlier code.

Removing the check.

On Sun, Apr 19, 2015 at 9:19 PM, Scott Feldman notifications@github.com
wrote:

> In hw/net/rocker/rocker_world.c
> https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28663372:
> 
> > @@ -97,10 +108,8 @@ enum rocker_world_type world_type(World *world)
> > 
> >  const char *world_name(World *world)
> >  {
> > -    switch (world->type) {
> > -    case ROCKER_WORLD_TYPE_OF_DPA:
> > -        return "OF_DPA";
> > -    default:
> > -        return "unknown";
> > -    if (world) {
> 
> Shouldn't need to check if world is non-null here. If NULL was passed in,
> that's a bug in the calling code and I think we'd rather have the device
> crash (so the bug can get fixed) rather than returning something that sort
> of look like world is non-NULL.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pbhide/qemu-rocker/pull/1/files#r28663372.

scottfeldman commented on hw/net/rocker/rocker_world.c:
```diff
@@ -67,6 +75,9 @@ void world_free(World *world)
     if (world->ops->uninit) {
         world->ops->uninit(world);
     }
+    if (world->name) {
```
Seems world->name will always be non-NULL here, so no need to check for NULL.

pbhide replied:
yes. will remove the check.

On Sun, Apr 19, 2015 at 9:22 PM, Scott Feldman notifications@github.com
wrote:

> In hw/net/rocker/rocker_world.c
> https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28663420:
> 
> > @@ -67,6 +75,9 @@ void world_free(World *world)
> >      if (world->ops->uninit) {
> >          world->ops->uninit(world);
> >      }
> > -    if (world->name) {
> 
> Seems world->name will always be non-NULL here, so no need to check for
> NULL.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pbhide/qemu-rocker/pull/1/files#r28663420.

scottfeldman commented on qapi/rocker.json:
```diff
@@ -75,7 +75,8 @@
             'speed': 'uint32', 'duplex': 'RockerPortDuplex',
-            'autoneg': 'RockerPortAutoneg' } }
+            'autoneg': 'RockerPortAutoneg',
+            'world_name': 'str'} }
```
nit: space between 'str' and '}'

scottfeldman commented on Makefile:
```diff
@@ -290,7 +290,7 @@ clean:
 	rm -f config.mak op-i386.h opc-i386.h gen-op-i386.h op-arm.h opc-arm.h gen-op-arm.h
 	rm -f qemu-options.def
-	find . \( -name '*.l[oa]' -o -name '*.so' -o -name '*.dll' -o -name '*.mo' -o -name '*.[oda]' \) -type f -exec rm {} +
+	find . -path ./hw/net/rocker/p4lib -prune -o \( -name '*.l[oa]' -o -name '*.so' -o -name '*.dll' -o -name '*.mo' -o -name '*.[oda]' \) -type f -exec rm {} +
```
What does this do?  Doesn't seem like we want rocker stuff in the top-level Makefile

pbhide replied:
Make clean at the top level was removing all libraries in all directories.
Since P4libs are currently not built as part of emu build they should not
be removed. These are checked-in libraries until we find a way to compiler
P4 program as part of emu build.
I added the above so that make clean does not remove the P4 libs. Could not
find more elegant way.

On Sun, Apr 19, 2015 at 9:25 PM, Scott Feldman notifications@github.com
wrote:

> In Makefile
> https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28663482:
> 
> > @@ -290,7 +290,7 @@ clean:
> >  # avoid old build problems by removing potentially incorrect old files
> >     rm -f config.mak op-i386.h opc-i386.h gen-op-i386.h op-arm.h opc-arm.h gen-op-arm.h
> >     rm -f qemu-options.def
> > -   find . ( -name '_.l[oa]' -o -name '_.so' -o -name '_.dll' -o -name '_.mo' -o -name '*.[oda]' ) -type f -exec rm {} +
> > -   find . -path ./hw/net/rocker/p4lib -prune -o ( -name '_.l[oa]' -o -name '_.so' -o -name '_.dll' -o -name '_.mo' -o -name '*.[oda]' ) -type f -exec rm {} +
> 
> What does this do? Doesn't seem like we want rocker stuff in the top-level
> Makefile
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pbhide/qemu-rocker/pull/1/files#r28663482.

scottfeldman commented on hw/net/Makefile.objs:
```diff
@@ -37,5 +37,6 @@ obj-$(CONFIG_ETSEC) += fsl_etsec/etsec.o fsl_etsec/registers.o \
 common-obj-$(CONFIG_ROCKER) += rocker/rocker.o rocker/rocker_fp.o \
-                               rocker/rocker_desc.o rocker/rocker_world.o \
-                               rocker/rocker_of_dpa.o
+								rocker/rocker_desc.o rocker/rocker_world.o \
```
Strange tab issue here...not sure what changed

pbhide replied:
My bad.. I had tab-stop set to 4 spaces.

On Sun, Apr 19, 2015 at 9:27 PM, Scott Feldman notifications@github.com
wrote:

> In hw/net/Makefile.objs
> https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28663511:
> 
> > @@ -37,5 +37,6 @@ obj-$(CONFIG_ETSEC) += fsl_etsec/etsec.o fsl_etsec/registers.o \
> >             fsl_etsec/rings.o fsl_etsec/miim.o
> > 
> >  common-obj-$(CONFIG_ROCKER) += rocker/rocker.o rocker/rocker_fp.o \
> > -                               rocker/rocker_desc.o rocker/rocker_world.o \
> > -                               rocker/rocker_of_dpa.o
> > -                               rocker/rocker_desc.o rocker/rocker_world.o \
> 
> Strange tab issue here...not sure what changed
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pbhide/qemu-rocker/pull/1/files#r28663511.

scottfeldman commented on hw/net/rocker/README.p4:
```diff
@@ -0,0 +1,45 @@
+    p4lib/p4utils.a
+
+
+P4 behavioral model uses libJudy. Please install ligjudy-dev package.
```
spelling ligjudy-dev

scottfeldman commented on hw/net/rocker/README.p4:
```diff
@@ -0,0 +1,45 @@
+P4 support for rocker device:
+--------------------------------
+
+P4 program is used define a new prcessing pipeline. This program 
```
spelling

scottfeldman commented on hw/net/rocker/rocker.c:
```diff
@@ -452,20 +456,37 @@ static int cmd_consume(Rocker *r, DescInfo *info)
         break;
     default:
-        err = -ROCKER_EINVAL;
+        if (tlvs[ROCKER_TLV_CMD_WORLD]) {
```
Hmm, let's just add the P4 cmds to case statement, and then world is known.

pbhide replied:
Same reasoning as earlier comment.

On Sun, Apr 19, 2015 at 9:30 PM, Scott Feldman notifications@github.com
wrote:

> In hw/net/rocker/rocker.c
> https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28663579:
> 
> > ```
> >      err = cmd_set_port_settings(r, info_tlv);
> >      break;
> >  default:
> > ```
> > -        err = -ROCKER_EINVAL;
> > -        if (tlvs[ROCKER_TLV_CMD_WORLD]) {
> 
> Hmm, let's just add the P4 cmds to case statement, and then world is known.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pbhide/qemu-rocker/pull/1/files#r28663579.

scottfeldman commented on hw/net/rocker/rocker.c:
```diff
@@ -473,7 +494,7 @@ static void rocker_msix_irq(Rocker *r, unsigned vector)
     PCIDevice *dev = PCI_DEVICE(r);
 
-    DPRINTF("MSI-X notify request for vector %d\n", vector);
+    // DPRINTF("MSI-X notify request for vector %d\n", vector);
```
leave uncommented

scottfeldman commented on hw/net/rocker/rocker.c:
```diff
@@ -997,9 +1018,9 @@ static const char *rocker_reg_name(void *opaque, hwaddr addr)
-    DPRINTF("Write %s addr " TARGET_FMT_plx
-            ", size %u, val " TARGET_FMT_plx "\n",
-            rocker_reg_name(opaque, addr), addr, size, val);
+    // DPRINTF("Write %s addr " TARGET_FMT_plx
```
leave uncommented

scottfeldman commented on hw/net/rocker/rocker_fp.c:
```diff
@@ -119,6 +120,7 @@ int fp_port_eg(FpPort *port, const struct iovec *iov, int iovcnt)
     NetClientState *nc = qemu_get_queue(port->nic);
 
     if (port->enabled) {
+	    DPRINTF("Send to port %d\n", port->pport);
```
Might get too chatty?

scottfeldman commented on hw/net/rocker/rocker_world.h:
```diff
@@ -56,5 +59,6 @@ enum rocker_world_type world_type(World *world);
 const char *world_name(World *world);
 
 World *rocker_get_world(Rocker *r, enum rocker_world_type type);
+World *rocker_world_from_pport(int pport);
```
Is this used anywhere?

pbhide replied:
not anymore. will remove.

On Sun, Apr 19, 2015 at 9:35 PM, Scott Feldman notifications@github.com
wrote:

> In hw/net/rocker/rocker_world.h
> https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28663678:
> 
> > @@ -56,5 +59,6 @@ enum rocker_world_type world_type(World *world);
> >  const char *world_name(World *world);
> > 
> >  World *rocker_get_world(Rocker *r, enum rocker_world_type type);
> > +World *rocker_world_from_pport(int pport);
> 
> Is this used anywhere?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pbhide/qemu-rocker/pull/1/files#r28663678.

scottfeldman commented on Makefile.target:
```diff
@@ -135,7 +135,7 @@ obj-y += memory.o savevm.o cputlb.o
 obj-y += dump.o
 obj-$(call lnot,$(CONFIG_ROCKER)) += hw/net/rocker/qmp-norocker.o
-LIBS+=$(libs_softmmu)
+LIBS+=$(libs_softmmu) -lJudy
```
Is there something already like libjudy in the qemu build tree?  libjudy looks nice but I can see a road-block trying to get it included for upstream qemu.

pbhide replied:
libjudy is used in the P4 behavioral model. That might be a big change for
us to migrate to something else like GHash.
Do you think it can result in "no-go"?

On Sun, Apr 19, 2015 at 9:55 PM, Scott Feldman notifications@github.com
wrote:

> In Makefile.target
> https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28664124:
> 
> > @@ -135,7 +135,7 @@ obj-y += memory.o savevm.o cputlb.o
> >  obj-y += memory_mapping.o
> >  obj-y += dump.o
> >  obj-$(call lnot,$(CONFIG_ROCKER)) += hw/net/rocker/qmp-norocker.o
> > -LIBS+=$(libs_softmmu)
> > +LIBS+=$(libs_softmmu) -lJudy
> 
> Is there something already like libjudy in the qemu build tree? libjudy
> looks nice but I can see a road-block trying to get it included for
> upstream qemu.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pbhide/qemu-rocker/pull/1/files#r28664124.

scottfeldman replied:
If the answer to "can ghash be used here instead of libjudy" is yes, then you probably need to convert to ghash.  I would expect push-back from qemu community on adding libjudy and this is the first question they'll ask.

scottfeldman commented on hw/net/rocker/rocker_p4_l2l3_enums.h:
```diff
@@ -0,0 +1,41 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc. 
+
+Licensed under the Apache License, Version 2.0 (the "License");
```
Can the generated files be licensed under GPL like the rest of the files?

pbhide replied:
Its a generated file.. let me try to fix the tool that generates it.

On Sun, Apr 19, 2015 at 9:59 PM, Scott Feldman notifications@github.com
wrote:

> In hw/net/rocker/rocker_p4_l2l3_enums.h
> https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28664188:
> 
> > @@ -0,0 +1,41 @@
> > +/*
> > +Copyright 2013-present Barefoot Networks, Inc.
> > +
> > +Licensed under the Apache License, Version 2.0 (the "License");
> 
> Why is this one under a diff license?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pbhide/qemu-rocker/pull/1/files#r28664188.

scottfeldman commented on hw/net/rocker/rocker_p4_l2l3_features.h:
```diff
@@ -0,0 +1,21 @@
+#ifndef _P4_L2L3_FEATURES_H_
+#define _P4_L2L3_FEATURES_H_
+
+#define CPU_PORT  0xAA // b-model drops packet sent to ports 0, 257
```
// style comments aren't allowed for qemu...need to convert to /\* */ comments.

scottfeldman commented on hw/net/rocker/rocker_p4_l2l3_features.h:
```diff
@@ -0,0 +1,21 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc. 
```
Not sure about open-ended copyrights.  Past and current dates OK, but future?  I'm not sure...might raise a flag.

pbhide replied:
I had the same question. I think it has to be a specific year. I'll work
internally to get it fixed.

On Sun, Apr 19, 2015 at 10:03 PM, Scott Feldman notifications@github.com
wrote:

> In hw/net/rocker/rocker_p4_l2l3_features.h
> https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28664265:
> 
> > @@ -0,0 +1,21 @@
> > +/*
> > +Copyright 2013-present Barefoot Networks, Inc.
> 
> Not sure about open-ended copyrights. Past and current dates OK, but
> future? I'm not sure...might raise a flag.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pbhide/qemu-rocker/pull/1/files#r28664265.

scottfeldman commented on hw/net/rocker/rocker_p4_l2l3_lf.h:
```diff
@@ -0,0 +1,48 @@
+  unsigned short standard_metadata_ingress_port;
+} rocker_p4_l2l3_mac_learn_digest_digest_entry_t;
+
+// Should be able to cast this to pipe_flow_lrn_msg_t.
```
// comment

scottfeldman commented on hw/net/rocker/rocker_p4_l2l3_lf.h:
```diff
@@ -0,0 +1,48 @@
+  rocker_p4_l2l3_mac_learn_digest_digest_entry_t  *entries;
+} rocker_p4_l2l3_mac_learn_digest_digest_msg_t;
+
+// Should be able to cast this to pipe_flow_lrn_notify_cb.
```
// comment

scottfeldman commented on hw/net/rocker/rocker_p4_rmt.c:
```diff
@@ -0,0 +1,216 @@
+        return 0;
+    }
+    p4_rmt->process_pkt(port, buf, pkt_len);
+    return iov_size(iov, iovcnt);
```
return pkt_len?

scottfeldman replied:
Where does buf get freed?

pbhide replied:
By the model code. (library)

On Sun, Apr 19, 2015 at 10:33 PM, Scott Feldman notifications@github.com
wrote:

> In hw/net/rocker/rocker_p4_rmt.c
> https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28664806:
> 
> > -                            const struct iovec *iov, int iovcnt)
> >   +{
> > -    int pkt_len = iov_size(iov, iovcnt);
> > -    unsigned int port = pport;  // model uses pports (1-N)
> > -    struct p4_rmt_world *p4_rmt = world_private(world);
> >   +
> > -    // ingress pipeline
> > -    DPRINTF("p4_rmt_ingress port %d, len %d .... \n", port, pkt_len);
> >   +
> > -    // convert iov to a pkt buffer
> > -    unsigned char *buf = rocker_p4_iov_to_buffer(iov, iovcnt);
> > -    if (buf == NULL) {
> > -        return 0;
> > -    }
> > -    p4_rmt->process_pkt(port, buf, pkt_len);
> > -    return iov_size(iov, iovcnt);
> 
> Where does buf get freed?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pbhide/qemu-rocker/pull/1/files#r28664806.

scottfeldman commented on hw/net/rocker/rocker_p4_rmt.c:
```diff
@@ -0,0 +1,216 @@
+    DPRINTF("p4_rmt_ingress port %d, len %d .... \n", port, pkt_len);
+
+    // convert iov to a pkt buffer
+    unsigned char *buf = rocker_p4_iov_to_buffer(iov, iovcnt);
```
I think in rocker (and this may be a qemu style thing) we tried to keep variable declarations before any code.

scottfeldman commented on hw/net/rocker/rocker_p4_rmt.c:
```diff
@@ -0,0 +1,216 @@
+    // ingress pipeline
+    DPRINTF("p4_rmt_ingress port %d, len %d .... \n", port, pkt_len);
+
+    // convert iov to a pkt buffer
```
// comments need to be converted to /\* comments */

scottfeldman commented on hw/net/rocker/rocker_p4_rmt.c:
```diff
@@ -0,0 +1,216 @@
+
+    len = iov_size(iov, iovcnt);
+    DPRINTF("convert %d iovs to a buffer of len %d\n", iovcnt, len);
+    buf = malloc(len);
```
Use g_malloc(), g_free()

pbhide replied:
This was done on purpose.. (But I need to fix it) as the P4 behavior model
frees the packet buffer using "free" not g_free.
But the model code does not allocate the buffer.. I'll abstract it by
creating pkt_buf_alloc() function that will be provided by the model
library.

On Sun, Apr 19, 2015 at 10:26 PM, Scott Feldman notifications@github.com
wrote:

> In hw/net/rocker/rocker_p4_rmt.c
> https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28664658:
> 
> > +unsigned char *rocker_p4_iov_to_buffer(const struct iovec *iov, int iovcnt);
> > +static int rocker_p4_rmt_table_cmd (struct p4_rmt_world *p4_rmt,
> > -                                    struct desc_info *info,
> > -                                    char *buf, unsigned short cmd,
> > -                                    RockerTlv **tlvs);
> >   +
> >   +unsigned char *
> >   +rocker_p4_iov_to_buffer(const struct iovec *iov, int iovcnt)
> >   +{
> > -    // convert iov to a single contiguous buffer for p4 model
> > -    int len = 0;
> > -    unsigned char *buf = NULL;
> >   +
> > -    len = iov_size(iov, iovcnt);
> > -    DPRINTF("convert %d iovs to a buffer of len %d\n", iovcnt, len);
> > -    buf = malloc(len);
> 
> Use g_malloc(), g_free()
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pbhide/qemu-rocker/pull/1/files#r28664658.

scottfeldman replied:
ok, cool

scottfeldman commented on hw/net/rocker/rocker_p4_rmt.c:
```diff
@@ -0,0 +1,216 @@
+    if (buf == NULL) {
+        return 0;
+    }
+    p4_rmt->process_pkt(port, buf, pkt_len);
```
not checking return code

pbhide replied:
Actually the process_pkt (currently) always returns 0. Also was not sure
what to return on the error.

How is the return value of the  NetClientInfo->receive_iov used by its
caller?
Looks like the receive_iov should return the bytes consumed ?? Can I return
0 if packet could not be processed for some reason.. Need to check.

I'll add the check.

On Sun, Apr 19, 2015 at 10:27 PM, Scott Feldman notifications@github.com
wrote:

> In hw/net/rocker/rocker_p4_rmt.c
> https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28664705:
> 
> > +ssize_t rocker_p4_rmt_ig (World *world, unsigned int pport,
> > -                            const struct iovec *iov, int iovcnt)
> >   +{
> > -    int pkt_len = iov_size(iov, iovcnt);
> > -    unsigned int port = pport;  // model uses pports (1-N)
> > -    struct p4_rmt_world *p4_rmt = world_private(world);
> >   +
> > -    // ingress pipeline
> > -    DPRINTF("p4_rmt_ingress port %d, len %d .... \n", port, pkt_len);
> >   +
> > -    // convert iov to a pkt buffer
> > -    unsigned char *buf = rocker_p4_iov_to_buffer(iov, iovcnt);
> > -    if (buf == NULL) {
> > -        return 0;
> > -    }
> > -    p4_rmt->process_pkt(port, buf, pkt_len);
> 
> not checking return code
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pbhide/qemu-rocker/pull/1/files#r28664705.

scottfeldman commented on hw/net/rocker/rocker_p4_rmt.c:
```diff
@@ -0,0 +1,216 @@
+{
+    unsigned int    table_id;
+    unsigned char   *entry;
+    if (tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ID] == NULL ||
```
blank line after declarations but before fr executable code.

scottfeldman commented on hw/net/rocker/rocker_p4_rmt.c:
```diff
@@ -0,0 +1,216 @@
+    rocker = p4_rmt->rocker;
+    w = p4_rmt->world;
+
+    fp_port_from_pport((unsigned int)ig_port1, &fp_ig_port);
```
where is fp_ig_port used?

scottfeldman commented on hw/net/rocker/rocker_p4_rmt.c:
```diff
@@ -0,0 +1,216 @@
+        case ROCKER_TLV_CMD_TYPE_P4_RMT_TABLE_ENTRY_ADD:
+            // check if fn is provided
+            if (p4_rmt->table_ops[table_id].add) {
+                p4_rmt->table_ops[table_id].add(entry);
```
Hmm...it looks like entry is a binary blob.  How is it packed?  Are there un-aligned access issues?  How is it versioned?  How is big/little/host/network byte order handled?  It seems we want TLV-style encode/decode of the entry fields to answer all the questions.

pbhide replied:
Yes it is a blob. One good thing is that, the code to create and consume
the entry  is generated by the p4 tool-chain so bit packing and endian-ness
is taken care of by the generated code. E.g. rocker_p4_l2l3_pd.c which has
entry build functions.

On Sun, Apr 19, 2015 at 10:48 PM, Scott Feldman notifications@github.com
wrote:

> In hw/net/rocker/rocker_p4_rmt.c
> https://github.com/pbhide/qemu-rocker/pull/1#discussion_r28665142:
> 
> > +
> > -    table_id = rocker_tlv_get_u32(tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ID]);
> > -    entry = rocker_tlv_data(tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ENTRY]);
> >   +
> > -    DPRINTF("p4_rmt_table_cmd for table %d\n", table_id);
> >   +
> > -    if (!p4_rmt->is_table_valid(table_id)) {
> > -        DPRINTF("p4_rmt_table_cmd for invalide table %d\n", table_id);
> > -        return 0;
> > -    }
> >   +
> > -    switch (cmd) {
> > -        case ROCKER_TLV_CMD_TYPE_P4_RMT_TABLE_ENTRY_ADD:
> > -            // check if fn is provided
> > -            if (p4_rmt->table_ops[table_id].add) {
> > -                p4_rmt->table_ops[table_id].add(entry);
> 
> Hmm...it looks like entry is a binary blob. How is it packed? Are there
> un-aligned access issues? How is it versioned? How is
> big/little/host/network byte order handled? It seems we want TLV-style
> encode/decode of the entry fields to answer all the questions.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/pbhide/qemu-rocker/pull/1/files#r28665142.

scottfeldman replied:
If it's generated code, then can a TLV format be generated?  I know that brings rocker's TLV packing into P4 tools, which is probably not what you want.  I think a blob is not going to be the correct answer, but I'm not sure where to slice this.  This is something we should revisit after all the other review work is done.

Commit:
- Fixed XXX comments
- Added code to handle mod and delete (still needs work in P4 model)
- Fixed the default action handling to include action data

Commit:
Re-enable the PCI ring and MSI printfs.. keep #if 1 around it to easily locate them. These printfs flood the output during debug.

Commit:
QEMU with new P4 program - rocker_p4.
Both programs, rocker_p4 and rocker_p4_l2l3 are compiled into the image except,
l2l3 model library is not linked and l2l3 table interface is replaced with a stub.
i.e. only one program can be used at a time.

Commit:
Code used for P4 workshop demo
